geopandas/tools/make_grid.py

```diff
@@ -0,0 +1,387 @@
+        such that a corner points upwards.
+
+    Returns
+    -------
+    GeoSeries
+        The returned GeoSeries contains the grid-cell centers, corners, or
+        polygons.
+
+    Examples
+    --------
+    >>> import geopandas
+    >>> world = geopandas.read_file(
+    ...     geopandas.datasets.get_path('naturalearth_lowres'))
+    >>> uruguay = world[world["name"] == "Uruguay"]
+    >>> sq_grid = geopandas.make_grid(uruguay,3)
+    >>> sq_grid
+    0    POLYGON ((-58.42707 -34.95265, -55.42707 -34.9...
+    1    POLYGON ((-58.42707 -31.95265, -55.42707 -31.9...
+    2    POLYGON ((-55.42707 -34.95265, -52.42707 -34.9...
+    3    POLYGON ((-55.42707 -31.95265, -52.42707 -31.9...
+    """
+
+    # Run basic checks
+    _basic_checks(input_geometry, cell_size, offset, what, cell_type, intersect)
+
+    output_grid = None
+
+    if isinstance(input_geometry, (GeoDataFrame, GeoSeries)):
+        bounds = np.array(input_geometry.total_bounds)
+    else:
+        bounds = np.array(input_geometry.bounds)
+
+    x_dist = bounds[2] - bounds[0] - offset[0]
+    y_dist = bounds[3] - bounds[1] - offset[1]
+
+    grid_origin_x = bounds[0] + offset[0]
+    grid_origin_y = bounds[1] + offset[1]
+
+    if cell_type == "square":
+        # Set corner coordinates of square grid. Grid always ends at the right/upper
+        # edge of the bounding box; also if an offset is defined.
+        x_coords_corn = np.arange(
+            bounds[0] + offset[0], bounds[2] + cell_size, cell_size
+        )
+        y_coords_corn = np.arange(
+            bounds[1] + offset[1], bounds[3] + cell_size, cell_size
+        )
+        xv, yv = np.meshgrid(x_coords_corn, y_coords_corn)
+
+        if what == "corners":
+            sq_corners_np = np.array([xv, yv]).T.reshape(-1, 2)
+            output_grid = points_from_xy(sq_corners_np[:, 0], sq_corners_np[:, 1])
+
+        elif what == "centers":
+            sq_centers_np = (
+                np.array([xv[:-1, :-1], yv[:-1, :-1]]).T.reshape(-1, 2) + cell_size / 2
+            )
+
+            output_grid = points_from_xy(sq_centers_np[:, 0], sq_centers_np[:, 1])
+
+        elif what == "polygons":
+            # Extracting corners of all square-grid cells.
+            bt_left_corners = np.array([xv[:-1, :-1], yv[:-1, :-1]]).T.reshape(-1, 1, 2)
+            bt_right_corners = np.array([xv[1:, 1:], yv[:-1, :-1]]).T.reshape(-1, 1, 2)
+            top_right_corners = np.array([xv[1:, 1:], yv[1:, 1:]]).T.reshape(-1, 1, 2)
+            top_left_corners = np.array([xv[:-1, :-1], yv[1:, 1:]]).T.reshape(-1, 1, 2)
+
+            sq_coords = np.concatenate(
+                (
+                    bt_left_corners,
+                    bt_right_corners,
+                    top_right_corners,
+                    top_left_corners,
+                ),
+                axis=1,
+            )
+            sq_polygons_np = np.array([Polygon(sq_set) for sq_set in sq_coords])
```

Comment:
Can you use the vectorized creation of polygons from arrays instead of this loop? Using https://shapely.readthedocs.io/en/stable/reference/shapely.polygons.html#shapely.polygons? With a possible fallback to this for older shapely.

Reply:
Brief follow up question on this one: 
GitHub tests fail for older versions of shapely because I imported `polygons` at the beginning of the py file:
`from shapely import polygons`
Later, when calling polygons i check for the shapely version:
```
if compat.SHAPELY_GE_20:
   sq_polygons = polygons(sq_coords)
else:
   sq_polygons = np.array([Polygon(sq_set) for sq_set in sq_coords])
```
How would you resolve this elegantly? Import shapely as a whole at the beginning or add an if statement to the import section?

Reply:
You can do just 
``` py
import shapely
from shapely.geometry import Polygon
```
on top and then do

```py
if compat.SHAPELY_GE_20:
   sq_polygons = shapely.polygons(sq_coords)
else:
   sq_polygons = np.array([Polygon(sq_set) for sq_set in sq_coords])
```

Reply:
Thanks! Ok, so it's not a concern to import the whole shapely package.